### PR TITLE
feat(cf): use stage timeouts for service broker operations

### PR DESCRIPTION
**Pull request by stuart-pollock**
spinnaker/spinnaker#3637

Some calls to CAPI require running in an asynchronous mode and must be
polled for their status.  These calls may effectively run forever
(for 14 days) within clouddriver because they may rely on orca stopping
such tasks with its own timeout mechanism.

- Also rename a function for consistency

### CODE REVIEW

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/client/ServiceInstances.java [hunk 1]
```diff
@@ -282,11 +279,11 @@ public void createUserProvidedServiceInstance(String newUserProvidedServiceInsta
     servicePolling.accept(guid, type);
   }
 
-  // package-level for testing visibility
+  @VisibleForTesting
   void pollServiceInstanceStatus(String serviceInstanceName, String guid, LastOperation.Type type, Duration timeout) {
     RetryConfig retryConfig = RetryConfig.custom()
       .waitDuration(pollingInterval)
-      .maxAttempts((int) (Math.max(timeout.toMillis(), defaultPollingTimeout.toMillis()) / pollingInterval.toMillis()))
+      .maxAttempts((int) (Math.max(timeout.toMillis(), pollingTimeout.toMillis()) / pollingInterval.toMillis()))
```
Comment (jkschneider) [inline clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/client/ServiceInstances.java]: Is there only one call site using `pollingTimeout` as the `timeout` argument, making this `max` comparison extraneous? Is it called elsewhere?

Comment (stuart-pollock) [reply clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/client/ServiceInstances.java]: 👍 Good catch

Comment (jkschneider) [reply clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/client/ServiceInstances.java]: Maybe I'm misreading, but it seems to me that now we've got:
`.maxAttempts((int) (timeout.toMillis() / pollingInterval.toMillis()))` which should always evaluate to 1 and `.waitDuration(pollingInterval)` which will always wait 14 days?

Comment (stuart-pollock) [reply clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/client/ServiceInstances.java]: I believe that we discussed this offline:  we need to make these calls to CAPI asynchronous so we need to poll.  We decided to use Spinnaker's built-in timeout to cancel the task when its timeout is reached.  The reason it is not inlined is for testing purposes only.

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/DeployCloudFoundryServerGroupAtomicOperationConverter.java [hunk 2]
```diff
@@ -21,7 +21,7 @@
 import com.fasterxml.jackson.databind.DeserializationFeature;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.databind.PropertyNamingStrategy;
-import com.google.common.annotations.VisibleForTesting;
+import com.google.api.client.repackaged.com.google.common.annotations.VisibleForTesting;
```
Comment (jkschneider) [inline clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/DeployCloudFoundryServerGroupAtomicOperationConverter.java]: Woops, wrong import.

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/DeployCloudFoundryServiceAtomicOperationConverter.java [hunk 3]
```diff
@@ -20,13 +20,13 @@
 import com.fasterxml.jackson.databind.DeserializationFeature;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.databind.PropertyNamingStrategy;
-import com.google.common.annotations.VisibleForTesting;
+import com.google.api.client.repackaged.com.google.common.annotations.VisibleForTesting;
```
Comment (jkschneider) [inline clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/DeployCloudFoundryServiceAtomicOperationConverter.java]: Wrong import.

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/description/AbstractCloudFoundryDescription.java [hunk 4]
```diff
@@ -21,7 +21,7 @@
 import lombok.Data;
 
 @Data
-public abstract class AbstractCloudFoundryDescription {
+abstract class AbstractCloudFoundryDescription {
```
Comment (jkschneider) [inline clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/description/AbstractCloudFoundryDescription.java]: No need to reduce visibility here?

File: clouddriver-cloudfoundry/src/test/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/ops/DeployCloudFoundryServiceAtomicOperationTest.java [hunk 5]
```diff
@@ -16,14 +16,25 @@
...
 
+import java.util.List;
+
+import static com.netflix.spinnaker.clouddriver.cloudfoundry.client.model.v2.LastOperation.State.IN_PROGRESS;
+import static com.netflix.spinnaker.clouddriver.cloudfoundry.client.model.v2.LastOperation.State.SUCCEEDED;
+import static com.netflix.spinnaker.clouddriver.cloudfoundry.client.model.v2.LastOperation.Type.CREATE;
+import static com.netflix.spinnaker.clouddriver.cloudfoundry.client.model.v2.LastOperation.Type.UPDATE;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.atIndex;
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
Comment (jkschneider) [inline clouddriver-cloudfoundry/src/test/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/ops/DeployCloudFoundryServiceAtomicOperationTest.java]: For consistency, can we use assertj throughout this test?

File: clouddriver-cloudfoundry/src/test/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/ops/DestroyCloudFoundryServiceAtomicOperationTest.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+import org.junit.jupiter.api.Test;
+
+import java.util.List;
+
+import static com.netflix.spinnaker.clouddriver.cloudfoundry.client.model.v2.LastOperation.State.IN_PROGRESS;
+import static com.netflix.spinnaker.clouddriver.cloudfoundry.client.model.v2.LastOperation.State.NOT_FOUND;
+import static com.netflix.spinnaker.clouddriver.cloudfoundry.client.model.v2.LastOperation.Type.DELETE;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.atIndex;
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
Comment (jkschneider) [inline clouddriver-cloudfoundry/src/test/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/ops/DestroyCloudFoundryServiceAtomicOperationTest.java]: Same, let's try to use assertj consistently.

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/DeployCloudFoundryServerGroupAtomicOperationConverter.java (hunk 2 above)
Comment (stuart-pollock) [reply clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/DeployCloudFoundryServerGroupAtomicOperationConverter.java]: 👍

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/DeployCloudFoundryServiceAtomicOperationConverter.java (hunk 3 above)
Comment (stuart-pollock) [reply clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/DeployCloudFoundryServiceAtomicOperationConverter.java]: 👍

File: clouddriver-cloudfoundry/src/test/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/ops/DestroyCloudFoundryServiceAtomicOperationTest.java (hunk 6 above)
Comment (stuart-pollock) [reply clouddriver-cloudfoundry/src/test/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/ops/DestroyCloudFoundryServiceAtomicOperationTest.java]: 👍

File: clouddriver-cloudfoundry/src/test/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/ops/DeployCloudFoundryServiceAtomicOperationTest.java (hunk 5 above)
Comment (stuart-pollock) [reply clouddriver-cloudfoundry/src/test/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/ops/DeployCloudFoundryServiceAtomicOperationTest.java]: 👍

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/description/AbstractCloudFoundryDescription.java (hunk 4 above)
Comment (stuart-pollock) [reply clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/description/AbstractCloudFoundryDescription.java]: 👍

File: clouddriver-cloudfoundry/src/test/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/ops/DestroyCloudFoundryServiceAtomicOperationTest.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+      .thenReturn(serviceInstanceResponse);
+
+    DestroyCloudFoundryServiceAtomicOperation op = new DestroyCloudFoundryServiceAtomicOperation(desc);
+
+    Task task = runOperation(op);
+    List<Object> resultObjects = task.getResultObjects();
+    assertThat(1).isEqualTo(resultObjects.size());
+    Object o = resultObjects.get(0);
+    assertThat(o).isInstanceOf(ServiceInstanceResponse.class);
+    ServiceInstanceResponse response = (ServiceInstanceResponse) o;
```
Comment (claymccoy) [inline clouddriver-cloudfoundry/src/test/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/ops/DestroyCloudFoundryServiceAtomicOperationTest.java]: I don't see much value in asserting the individual fields of the response you created above. Maybe an equals against the whole object if you want to assert that your response is making it into the right place. I prefer tests that just have one (or very few asserts). In this case I guess the important one is the history? That just indirectly lets you know what happened though.

Comment (stuart-pollock) [reply clouddriver-cloudfoundry/src/test/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/ops/DestroyCloudFoundryServiceAtomicOperationTest.java]: 👍

**Pull request merged**